core/kazoo_amqp/src/api/kapi_im.erl

```diff
@@ -11,137 +11,210 @@
+                                                            ,<<"Route-Type">>
+                                                            ,<<"System-ID">>
+                                                            ,<<"Target-Flags">>
+                                                            ,<<"Target-Properties">>
+                                                            ,<<"Transaction-ID">>
+                                                            ]}
+              ,{fun kapi_definition:set_values/2
+               ,[{<<"Route-Type">>, [<<"onnet">>, <<"offnet">>]}
+                 | kapi_definition:event_type_headers(Category, EventName)
+                ]
+               }
+              ,{fun kapi_definition:set_types/2
+               ,[{<<"Body">>, fun is_binary/1}
+                ,{<<"Custom-Vars">>, fun kz_json:is_json_object/1}
+                ,{<<"Endpoints">>, fun is_list/1}
+                ,{<<"From">>, fun is_binary/1}
+                ,{<<"Message-ID">>, fun is_binary/1}
+                ,{<<"Request">>, fun is_binary/1}
+                ,{<<"System-ID">>, fun is_binary/1}
+                ,{<<"To">>, fun is_binary/1}
+                ]}
+              ],
+    kapi_definition:setters(Setters).
+
+%%------------------------------------------------------------------------------
+%% @doc SMS MMS Inbound
+%% @end
+%%------------------------------------------------------------------------------
+-spec inbound(kz_term:api_terms()) -> kz_api:api_formatter_return().
+inbound(Req) ->
+    kapi_definition:build_message(Req, inbound_definition()).
 
 -spec inbound_v(kz_term:api_terms()) -> boolean().
-inbound_v(Prop) when is_list(Prop) ->
-    kz_api:validate(Prop, ?INBOUND_HEADERS, ?INBOUND_REQ_VALUES, ?INBOUND_TYPES);
-inbound_v(JObj) -> inbound_v(kz_json:to_proplist(JObj)).
-
--spec outbound(kz_term:api_terms()) -> {'ok', iolist()} |
-          {'error', string()}.
-outbound(Prop) when is_list(Prop) ->
-    case outbound_v(Prop) of
-        'true' -> kz_api:build_message(Prop, ?OUTBOUND_HEADERS, ?OPTIONAL_OUTBOUND_HEADERS);
-        'false' -> {'error', "Proplist failed validation for outbound"}
-    end;
-outbound(JObj) -> outbound(kz_json:to_proplist(JObj)).
+inbound_v(Req) ->
+    kapi_definition:validate(Req, inbound_definition()).
+
+-spec publish_inbound(kz_term:api_terms()) -> 'ok'.
+publish_inbound(Req) ->
+    publish_inbound(Req, []).
+
+-spec publish_inbound(kz_term:api_terms(), kz_term:proplist()) -> 'ok'.
+publish_inbound(Req, AMQPOptions) ->
+    Definition = inbound_definition(),
+    {'ok', Payload} = kz_api:prepare_api_payload(Req
+                                                ,kapi_definition:values(Definition)
+                                                ,kapi_definition:build_fun(Definition)
+                                                ),
+    kz_amqp_util:basic_publish(exchange_id(Req)
+                              ,(kapi_definition:binding(Definition))(Req)
```

Comment:
this code is a bit confusing. i know it return a `fun` but still confusing. maybe add `kapi_definition:binding/2`   that takes definition and api_terms and returns the routing key

Reply:
It is not possible because every module uses a different number of values for constructing routing keys.